lua/bufferline/options.lua

```diff
@@ -1,17 +1,26 @@
+local table_concat = table.concat
+
 local ERROR = vim.diagnostic.severity.ERROR --- @type integer
 local HINT = vim.diagnostic.severity.HINT --- @type integer
 local INFO = vim.diagnostic.severity.INFO --- @type integer
 local tbl_deep_extend = vim.tbl_deep_extend
 local WARN = vim.diagnostic.severity.WARN --- @type integer
 
+local utils = require'bufferline.utils'
+
+--- The prefix used for `utils.deprecate`
+local DEPRECATE_PREFIX = '\nThe barbar.nvim option '
+
+--- A cached value of `g:bufferline`
+local g_bufferline = {}
```

Comment:
`g:bufferline` is now cached. This massively increases performance, since we don't have to do `(vim.g.bufferline or {}).foo` on every `options.foo()`.

The `dictwatcher`s and `vim.g` watcher should take care of updating the cache, and since `setup` updates `vim.g`, that's covered too. I've tested it pretty extensively, and I've noted the only potential pitfalls under my comments about the `dictwatcheradd` changes.